wasm/lib/src/module.dart

```diff
@@ -454,8 +454,20 @@ class WasmGlobal {
 
   set value(dynamic val) {
     if (_mut == wasmerMutabilityConst) {
-      throw WasmError("Can't set value of const global: $this");
+      throw _WasmErrorImpl("Can't set value of const global: $this");
     }
     runtime.globalSet(_global, _type, val);
   }
 }
+
+class _WasmErrorImpl extends Error implements WasmError {
```

Comment:
Duplicate of the one in runtime.dart?

Reply:
These are distinct subclasses so that we don't have to introduce a dependency on a common WasmError subclass. Since they are private, users won't be able to refer to them in code, so them having the same name shouldn't be an issue.

Reply:
It's code duplication. That's why I had these declarations in wasm_error.dart. What's wrong with that approach?

Reply:
Are you implying that code duplication is bad? In many cases, yes, but in this case I would have to disagree.

The reasoning behind my stance is the following:

1) Both classes have a different purpose. The intention of one of the two errors is to represent an error that has occurred within module.dart and the other within runtime.dart. By having two separate types we make these intentions explicit in the type system. This will also make it easier to expose more precise error types i.e. see: https://github.com/lexaknyazev/wasm.dart/blob/a6c93afea4732c140f1f61f144795961c42c8613/wasm_interop/lib/wasm_interop.dart#L677-L710).

2) Analogously to how it is common to denormalize databases (i.e. introduce redundancy) to increase performance, we can trade having what appears to be duplicated code for better maintainability. (my definition of maintainability is rather abstract, but if you're familiar with graph theory, a system with more [bridges](https://en.wikipedia.org/wiki/Bridge_(graph_theory)) is more maintainable, because the more bridges there are, the easier it is to remove components out of a system i.e. increase the number of connected components. A single concrete Error type instantiated in both files affects that negatively)

3) I'm not saying it would be wrong to have a separate error class that is shared across the two files, but the benefits of having both of them be private outweigh the downside of having them at two places. (They are private so nobody outside of their respective files can ever depend on them, this is a huge plus in my opinion)

Reply:
> the benefits of having both of them be private outweigh the downside of having them at two places
If by private you mean private to users, you can just not show them in wasm.dart. If you mean private between the modules within this library, then I'm not convinced that's a benefit.

> Both classes have a different purpose.
The divide between module and runtime is a bit arbitrary (more to do with breaking up files that were getting too long, and less of a principled modularization). But in general the idea of having different exceptions for different purposes is good. If that's your goal though, then they should have different names, and we'll probably need to divide them more granularly in future.

Reply:
> If by private you mean private to users, you can just not show them in wasm.dart

Yes, the former.

If we were to only hide them from wasm.dart, then clients of the package would still be able to depend on them by importing e.g. `.../src/wasm_error.dart`. Hiding types by excluding them from the barrel file is an unsound convention that keeps the language simple. By having the classes be dart-private we can soundly guarantee that nobody can ever depend on them. (_except through hacky introspection facilities like dart:mirror which it looks like will soon be removed_).

A different way to put this could be in the words of [hyrum's law](https://www.hyrumslaw.com/). Not exporting a type and keeping it dart-public would leave it observable. But making it dart-private no outside user would be able to observe it.

I agree that it would be better to give them distinct names.

Reply:
Done